src/org/ojalgo/scalar/RationalNumber.java

```diff
@@ -177,50 +173,33 @@ public static RationalNumber valueOf(final double value) {
             return NEGATIVE_INFINITY;
         }
 
-        //        final long valBits = Double.doubleToLongBits(value);
-        //
-        //        final int sign = ((valBits >> 63) == 0 ? 1 : -1);
-        //        int exponent = (int) ((valBits >> 52) & 0x7ffL);
-        //        final long significand = (exponent == 0 ? (valBits & ((1L << 52) - 1)) << 1 : (valBits & ((1L << 52) - 1)) | (1L << 52));
-        //        exponent -= 1075;
-        //
-        //        if (exponent < 0) {
-        //            return RationalNumber.of(sign * significand, 1L << exponent);
-        //        } else {
-        //            return RationalNumber.of(sign * significand * (1L << exponent), 1L);
-        //        }
-
-        // Code below doesn't work for 0 and NaN - just check before
-
-        //        final long bits = Double.doubleToLongBits(value);
-        //
-        //        final long sign = bits >>> 63;
-        //        final long exponent = ((bits >>> 52) ^ (sign << 11)) - 1023;
-        //        final long fraction = bits << 12; // bits are "reversed" but that's not a problem
-        //
-        //        long a = 1L;
-        //        long b = 1L;
-        //
-        //        for (int i = 63; i >= 12; i--) {
-        //            a = (a * 2) + ((fraction >>> i) & 1);
-        //            b *= 2;
-        //        }
-        //
-        //        if (exponent > 0) {
-        //            a *= 1 << exponent;
-        //        } else {
-        //            b *= 1 << -exponent;
-        //        }
-        //
-        //        if (sign == 1) {
-        //            a *= -1;
-        //        }
-        //
-        //        // Here you have to simplify the fraction
-        //
-        //        System.out.println(a + "/" + b);
-
-        return RationalNumber.valueOf(BigDecimal.valueOf(value));
+        final long bits = Double.doubleToLongBits(value);
+
+        // Please refer to {@link Double#doubleToLongBits(long)} javadoc
+        final int s = ((bits >> 63) == 0) ? 1 : -1;
+        final int e = (int) ((bits >> 52) & 0x7ffL);
+        long m = (e == 0) ?
+                (bits & 0xfffffffffffffL) << 1 :
+                (bits & 0xfffffffffffffL) | 0x10000000000000L;
+        // Now we're looking for s * m * 2^{e - 1075}, 1075 being bias of 1023 plus 52 positions of binary fraction
+
+        int exponent = e - 1075;
+
+        if (exponent >= 0) {
+            // TODO: Check for MAX_BITS and report infinity or throw
+            return new RationalNumber(s * (m << exponent), 1L);
+        }
+
+        while ((m & 1) == 0 && exponent < 0) {
+            m >>= 1;
+            exponent++;
+        }
+        if (-exponent >= MAX_BITS) {
+            int shift = MAX_BITS - 1;
+            m >>= -exponent - shift;
+            return new RationalNumber(s * m, 1L << shift);
```

Comment:
That's the breaking change; if we leave `if (-exponent > MAX_BITS)` and shift by `-exponent`, then in case the exponent happens to be `-MAX_BITS` we use `1L << MAX_BITS` as a denominator, which is `-9223372036854775808`, so not exactly what we're looking for.

How exactly that affects the singular value decomposition is not quite clear to me. Why on Earth that could possibli work with rationals messed up is not too clear, either.

Reply:
By "breaking" I mean there's a noticeable correlation between this problem being fixed and the SV decomposition tests being broken. False positives in CI make me think that reality is trickier than "that's the change" - but it does seem that the probability of failure changes dramatically with this fix.

Reply:
Oh of course! Got it, mea culpa. Fix coming shortly.